compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -2647,6 +2649,76 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         span.is_desugaring(DesugaringKind::QuestionMark)
             && self.tcx.is_diagnostic_item(sym::From, trait_def_id)
     }
+
+    /// Structurally compares two types, modulo any inference variables.
+    ///
+    /// Returns `true` if two types are equal, or if one type is an inference variable compatible
+    /// with the other type. A TyVar inference type is compatible with any type, and an IntVar or
+    /// FloatVar inference type are compatible with themselves or their concrete types (Int and
+    /// Float types, respectively). When comparing two ADTs, these rules apply recursively.
+    pub fn same_type_modulo_infer(&self, a: Ty<'tcx>, b: Ty<'tcx>) -> bool {
+        let (a, b) = self.resolve_vars_if_possible((a, b));
+        match (a.kind(), b.kind()) {
+            (&ty::Adt(def_a, substs_a), &ty::Adt(def_b, substs_b)) => {
+                if def_a != def_b {
+                    return false;
+                }
+
+                substs_a
+                    .types()
+                    .zip(substs_b.types())
+                    .all(|(a, b)| self.same_type_modulo_infer(a, b))
+            }
+            (&ty::FnDef(did_a, substs_a), &ty::FnDef(did_b, substs_b)) => {
+                if did_a != did_b {
+                    return false;
+                }
+
+                substs_a
+                    .types()
+                    .zip(substs_b.types())
+                    .all(|(a, b)| self.same_type_modulo_infer(a, b))
+            }
```

Comment:
```suggestion
            (&ty::Adt(def_a, substs_a), &ty::Adt(def_b, substs_b))
            | (&ty::FnDef(did_a, substs_a), &ty::FnDef(did_b, substs_b)) => {
                if did_a != did_b {
                    return false;
                }

                substs_a
                    .types()
                    .zip(substs_b.types())
                    .all(|(a, b)| self.same_type_modulo_infer(a, b))
            }
```

Reply:
`def_a` is type `AdtDef` and `did_a` is type `DefId`.

Reply:
Oops didn't see that sorry

Reply:
r=me after fmt